src/garage/replay_buffer/path_buffer.py

```diff
@@ -0,0 +1,132 @@
+"""A replay buffer that efficiently stores and can sample whole paths."""
+import collections
+
+import numpy as np
+
+
+class PathBuffer:
+    """A replay buffer that stores and can sample whole paths.
+
+    This buffer only stores valid steps, and doesn't require paths to
+    have a maximum length.
+    Args:
+        capacity_in_steps (int): total memory allocated for the buffer
+    """
+
+    def __init__(self, capacity_in_transitions):
+        self._capacity = capacity_in_transitions
+        self._transitions_stored = 0
+        self._first_idx_of_next_path = 0
+        # Each path in the buffer has a tuple of two ranges in
+        # self._path_segments. If the path is stored in a single contiguous
+        # region of the buffer, the second range will be range(0, 0).
+        # The "left" side of the deque contains the oldest path.
+        self._path_segments = collections.deque()
+        self._buffer = {}
+
+    def add_path(self, path):
+        """Add a path to the buffer.
+
+        Args:
+            path(dict): A dict of array of shape (path_len, flat_dim)
+        """
+        for key, buf_arr in self._buffer.items():
+            path_array = path.get(key, None)
+            if path_array is None:
+                raise ValueError('Key {} missing from path.'.format(key))
+            elif (len(path_array.shape) != 2
+                  or path_array.shape[1] != buf_arr.shape[1]):
+                raise ValueError('Array {} has wrong shape.'.format(key))
+        path_len = self._get_path_length(path)
+        first_seg, second_seg = self._next_path_segments(path_len)
+        # Remove paths which will overlap with this one.
+        while (self._path_segments and self._segments_overlap(
+                first_seg, self._path_segments[0][0])):
+            self._path_segments.popleft()
+        while (self._path_segments and self._segments_overlap(
+                second_seg, self._path_segments[0][0])):
+            self._path_segments.popleft()
+        self._path_segments.append((first_seg, second_seg))
+        for key, array in path.items():
+            buf_arr = self._get_or_allocate_key(key, array)
```

Comment:
This logic seems similar to using a collections.defaultdict data structure. Maybe that is best used here.

Reply:
The problem is that there's no type we can set for the `defaultdict` that will work, since the array's shape depends on what we're trying to insert into it.